### PR TITLE
C++: Use GVN in AV Rule 79

### PR DESCRIPTION
**Pull request by MathiasVP**
Fixes https://github.com/github/codeql/issues/5412.

Consider [this allocation](https://lgtm.com/projects/g/coin3d/coin/snapshot/f91ad7f35d0a659a39e5a1d30309b2db7ac0517a/files/src/shadows/SoShadowGroup.cpp?sort=name&dir=ASC&mode=heatmap#L391) which is being freed [here](https://lgtm.com/projects/g/coin3d/coin/snapshot/f91ad7f35d0a659a39e5a1d30309b2db7ac0517a/files/src/shadows/SoShadowGroup.cpp?sort=name&dir=ASC&mode=heatmap#L498) by a call to `unref`. At the very bottom of `unref` is this piece of code that does the actual free'ing:
```cpp
if (refcount == 0) {
  SoBase * base = const_cast<SoBase *>(this);
  base->destroy();
}
```
Prior to this PR, we did not recognize that the call to `base->destroy()` was free'ing `this`. This PR uses `GVN` to make this connection between `this` and `base`.

Performance looks fine on `coin3d/coin` and Linux. It doesn't seem to have a big impact on the usual ~80 projects on LGTM. I added the user of the issue's project to show that the change _does_ have an impact on their specific codebase: https://lgtm.com/query/1742353848607434356/

<s>I suppose we could make a similar extension to the other recursive case in `exprReleases`, but for simplicity, I chose to keep that one as-is.</s> The change was so simple that I extended the other case as well in 2abf4c068f79785126a617a40b048c064b786f36. The only change on the 80 projects is still just the original user's project: https://lgtm.com/query/8218180625489198233/

### CODE REVIEW

File: cpp/ql/src/jsf/4.10 Classes/AV Rule 79.ql [hunk 1]
```diff
@@ -110,7 +111,8 @@ private predicate exprReleases(Expr e, Expr released, string kind) {
     ) and
     e.(FunctionCall).getQualifier() = exprOrDereference(released) and
     innerThis.getEnclosingFunction() = f and
-    exprReleases(_, innerThis, kind)
+    globalValueNumber(innerThis) = globalValueNumber(likeInnerThis) and
+    exprReleases(_, likeInnerThis, kind)
```
Comment (rdmarsh2) [inline cpp/ql/src/jsf/4.10 Classes/AV Rule 79.ql:115]: You could avoid declaring `likeInnerThis` by using `globalValueNumber(innerThis).getAnExpr()`.

**Review by rdmarsh2 — COMMENTED** [review]
LGTM otherwise, but avoiding the variable might make it simple enough to change the recursive case.

File: cpp/ql/src/jsf/4.10 Classes/AV Rule 79.ql (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/jsf/4.10 Classes/AV Rule 79.ql]: Good point! Fixed in 2abf4c068f79785126a617a40b048c064b786f36.

**Review by MathiasVP — COMMENTED** [review]
> LGTM otherwise, but avoiding the variable might make it simple enough to change the recursive case.

Yeah, I extended the PR to also support the other recursive case in 2abf4c068f79785126a617a40b048c064b786f36. Performance looks unchanged.

**Review by geoffw0 — APPROVED** [review]
:+1:

**Pull request merged**